marathon/src/main/scala/io/buoyant/marathon/v2/AppIdNamer.scala

```diff
@@ -0,0 +1,119 @@
+package io.buoyant.marathon.v2
+
+import com.twitter.finagle.{Addr, Name, Namer, NameTree, Path}
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.util._
+import java.net.SocketAddress
+
+object AppIdNamer {
+  object Closed extends Throwable
+  val log = Api.log
+}
+
+class AppIdNamer(
+  api: Api,
+  prefix: Path,
+  ttl: Duration,
+  timer: Timer = DefaultTimer.twitter
+) extends Namer {
+
+  import AppIdNamer._
+
+  private[this] implicit val _timer = timer
+
+  /**
+   * Accepts names in the form:
+   *   /<app-id>/residual/path
+   *
+   * and attempts to bind an Addr by resolving named endpoint from the
+   * Marathon master.
+   */
+  def lookup(path: Path): Activity[NameTree[Name]] = path match {
+    case Path.Utf8(app, _*) =>
+      val id = prefix ++ Path.Utf8(app)
+      val residual = path.drop(1)
+      // Marathon appends a leading '/' to app names that don't have it.
+      val appId = if (app(0) == '/') {
+        app
+      } else {
+        s"/$app"
+      }
+      log.debug("marathon lookup: %s %s %s", id.show, path.show, appId)
+      appsActivity.map {
+        case apps if apps(appId) =>
+          val addr = watchAddr(appId)
+          log.debug("marathon app found: (%s) appId: %s addr: %s", apps.mkString(", "), appId, addr)
+          NameTree.Leaf(Name.Bound(addr, id, residual))
+        case apps =>
+          log.debug("marathon app not found: (%s) appId: %s", apps.mkString(", "), appId)
+          NameTree.Neg
+      }
+
+    case _ => Activity.value(NameTree.Neg)
+  }
+
+  private[this] val appsActivity: Activity[Api.AppIds] = {
+    val states = Var.async[Activity.State[Api.AppIds]](Activity.Pending) { state =>
+      def loop(): Future[Unit] =
+        api.getAppIds().transform {
+          case Return(apps) =>
+            state() = Activity.Ok(apps)
+            Future.sleep(ttl).before(loop())
+
+          case Throw(NonFatal(e)) =>
+            state() = Activity.Failed(e)
+            Future.sleep(ttl).before(loop())
+
+          case Throw(e) =>
+            state() = Activity.Failed(e)
+            Future.exception(e)
+        }
+
+      val work = loop()
+      Closable.make { deadline =>
+        work.raise(Closed)
+        Future.Unit
+      }
+    }
+    Activity(states)
+  }
+
+  private[this] var appMonitors: Map[String, Var[Addr]] = Map.empty
+  private[this] def watchAddr(app: String): Var[Addr] = synchronized {
```

Comment:
hmmm, this is a bit weird because the function both retrieves an address, and also initiates an ongoing watch as a side-effect. we could call it `getMonitoredAddrAndStartWatching`, but i'm inclined to just leave it as-is, to at least be explicit about the side-effect part.